### PR TITLE
Implement the ComplEx knowledge graph algorithm

### PR DESCRIPTION
**Pull request by huonw**
This implements the ComplEx knowledge graph algorithm from: [Complex Embeddings for Simple Link Prediction](http://jmlr.org/proceedings/papers/v48/trouillon16.pdf), Théo Trouillon, Johannes Welbl, Sebastian Riedel, Éric Gaussier and Guillaume Bouchard, ICML 2016.

This comes in three parts:

1. a `KGTripleGenerator` where the `flow` method takes a list of `(source, edge type, target)` triples (aka `(subject, relation, object)`) and yields them as batches of sequential integers to use to look-up in embedding layers
2. the `ComplEx` model which creates the appropriate embedding layers and applies the ComplEx scoring mechanism to them (which gives logit "probabilities" for a link)
3. some minor changes to `StellarGraph` and its supporting code, so that passing in an unknown node gets an appropriate error

This doesn't add a notebook that reproduces the paper, because the datasets used there are new to the library, and the MRR metrics used there are awkward to compute. This can/should be future work (https://github.com/stellargraph/stellargraph/issues/862). However it does add tests, include one that does simple and manual computation of the scores for links, to validate that it matches the paper's description.

Future work: There's a whole class of knowledge graph algorithms that have a similar approach (e.g. DistMult #755), just changing the details of the embeddings (basically real or complex, matrix or vector) and the scoring mechanism. The `KGTripleGenerator` class can be easily reused for them, and there's likely some commonalities that we can factor out of the `ComplEx` model to share code. In the interests of YAGNI and waiting to choose the right generalisation, this PR does not try to generalise the `ComplEx` model yet. The ComplEx paper has a good summary of these similar algorithms:

image

See: #756

### CODE REVIEW

File: stellargraph/mapper/knowledge_graph.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            rel_ilocs=rel_ilocs,
+            target_ilocs=target_ilocs,
+            batch_size=self.batch_size,
+            shuffle=shuffle,
+            negative_samples=negative_samples,
+            seed=seed,
+        )
+
+
+class KGTripleSequence(Sequence):
```
Comment (kieranricardo) [inline stellargraph/mapper/knowledge_graph.py]: Could this class be replaced with a tensorflow `Dataset`?

Something along the lines of:
`dataset = tf.data.Dataset.from_tensor_slices([source_ilocs, rel_ilocs, target_ilocs]).shuffle().batch(batch_size).map(negative_sampling_batch_function)` 

Would do the trick. This should be faster, allow multithreading, and would remove the boiler plate batch shuffling code.

File: stellargraph/layer/knowledge_graph.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shape):
+        self.built = True
+
+    def call(self, inputs):
+        s_re, s_im, r_re, r_im, o_re, o_im = inputs
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py]: Sorry about that - I'll make a new comment with suggestions below referring to the right bit of code

[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py]:
```suggestion
        r_re, r_im, s_re, s_im, o_re, o_im = inputs
```

For me, keeping the ordering consistent with what the `inner` function expects would improve the readability - but again its a minor thing :)

File: stellargraph/mapper/knowledge_graph.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self._global_lock = threading.Lock()
+
+    def _batch_sampler(self, batch_num):
+        self._global_lock.acquire()
+        try:
+            return self._batch_samplers[batch_num]
+        except IndexError:
+            new_samplers = batch_num - len(self._batch_samplers) + 1
+            seeds = self._global_rs.integers(2 ** 32, size=new_samplers)
+            self._batch_samplers.extend(np.random.default_rng(seed) for seed in seeds)
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/mapper/knowledge_graph.py]: Makes sense! it looks like `tf.data` suffers from the same drawbacks. 

I briefly tried a workaround similar to the approach here where you use a batch numbering system to generate random seeds but ran into issues with using tensors as random seeds.

This is less trivial than I thought and probably better suited to future work. 

Another approach would be to generate the random nodes/triples to change with a single core process and using multiprocessing for everything else:

`dataset.map(generate_random_numbers, num_parallel_calls=1).map(change_triples, num_parallel_calls=num_parallel_calls)`

The downside here is that the `generate_random_numbers` isn't parallel, but testing on my machine parallelising generating random numbers with tensorflow isn't sped up by using multiprocessing.

File: stellargraph/mapper/knowledge_graph.py (hunk 1 above)
Comment (kieranricardo) [reply stellargraph/mapper/knowledge_graph.py]: sorry typo above, "multithreading" should've been "c multiprocessing" - which I believe has less overhead than python multiprocessing.

File: stellargraph/layer/knowledge_graph.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        k (int): the dimension of the embedding (that is, a vector in C^k is learnt for each node
+            and each link type)
+
+        embedding_initializer (str or func, optional): The initialiser to use for the embeddings.
+
+        embedding_regularizer (str or func, optional): The regularizer to use for the embeddings.
+    """
+
+    def __init__(
+        self, generator, k, embedding_initializer=None, embedding_regularizer=None,
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py]: Yeah the `=None` behaviour here is annoying!

I think matching the paper by default is usually good option. :)

File: stellargraph/layer/knowledge_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/layer/knowledge_graph.py]: The structure is like this because `s, r, o` matches visualising this as `<node> --<edge type>-- <node>`, and it's how the paper phrases the scoring function `ɸ(s, r, o)` (sometimes at least), and seems to be a common way to phrase such knowledge graph computations (e.g. RESCAL, TransE and even NTN in the screenshot in the PR description). The `inner` function matches how the paper describes the inner products themselves:

image

I can see what you're getting at... but I don't know whether uniformity here is the right choice.

File: stellargraph/mapper/knowledge_graph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (huonw) [reply stellargraph/mapper/knowledge_graph.py]: > Another approach would be to generate the random nodes/triples to change with a single core process and using multiprocessing for everything else:

Interesting idea 👍 I'll leave it for future work, if that's ok: https://github.com/stellargraph/stellargraph/issues/876

File: stellargraph/mapper/knowledge_graph.py (hunk 1 above)
Comment (huonw) [reply stellargraph/mapper/knowledge_graph.py]: (As above, https://github.com/stellargraph/stellargraph/issues/876)

File: stellargraph/layer/knowledge_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/knowledge_graph.py]: ah I see what you mean, agreed better to be consistent with the equations in that case :)

File: stellargraph/mapper/knowledge_graph.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            negative_count = self.negative_samples * positive_count
+            assert len(s_iloc) == positive_count + negative_count
+
+            rng = self._batch_sampler(batch_num)
+
+            change_source = rng.randint(2, size=negative_count).astype(bool)
+            source_changes = change_source.sum()
+            new_nodes = rng.randint(self.max_node_iloc, size=negative_count)
+
+            s_iloc[positive_count:][change_source] = new_nodes[:source_changes]
```
Comment (kieranricardo) [inline stellargraph/mapper/knowledge_graph.py]: Could this be replaced with:

```suggestion
            s_iloc[positive_count:][:(negative_count // 2)] = new_nodes[:(negative_count // 2)]
```

And vice versa with the line below? This would avoid computing `change_source` which is roughly 30% of the time in this function (based on some rough timing) and should still be random enough

Comment (huonw) [reply stellargraph/mapper/knowledge_graph.py]: I'm nervous about doing this optimisation, without having any code/notebooks with which to validate that it's not degrading the models. I'll add a note in the code but not do it yet. How does that sound?

Comment (huonw) [reply stellargraph/mapper/knowledge_graph.py]: Ah, also, I found that `np.random(size=...) < 0.5` is about 2&times; faster than `np.randint(2, size=...)`.

Comment (kieranricardo) [reply stellargraph/mapper/knowledge_graph.py]: That's a fair point, I think that sounds reasonable.
